[9.0.0][Formulas] Make all CRUD-related hook callbacks pass physical indexes to the formula engine.

jansiegel opened the pull request:
### Context
Make all CRUD-related hook callbacks pass physical indexes to the formula engine.

<sup>[skip changelog]</sup>

### How has this been tested?
Added a test case.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. #8091 

### Affected project(s):
- [x] `handsontable`
- [ ] `@handsontable/angular`
- [ ] `@handsontable/react`
- [ ] `@handsontable/vue`

budnix reviewed (changes requested):
It seems that not all changes are covered with the tests. When I reverted the login from the following hooks:
 * `onBeforeCreateRow`
 * `onBeforeCreateCol`
 * `onBeforeRemoveRow`
 * `onBeforeRemoveCol`
 * `onAfterCreateRow`
 * `onAfterCreateCol`
 * `onAfterRemoveCol`

The tests still pass. Can I ask you to write the missing tests?

aninde reviewed (approved):
🥇

jansiegel commented:
> It seems that not all changes are covered with the tests. When I reverted the logic from the following hooks:
> 
> * `onBeforeCreateRow`
> * `onBeforeCreateCol`
> * `onBeforeRemoveRow`
> * `onBeforeRemoveCol`
> * `onAfterCreateRow`
> * `onAfterCreateCol`
> * `onAfterRemoveCol`
> 
> The tests still pass. Can I ask you to write the missing tests?

I split the test case into 8 separate cases - one covering each hook callback.

budnix commented on src/plugins/formulas/__tests__/formulas.spec.js:
```diff
@@ -2098,6 +2098,197 @@ describe('Formulas general', () => {
+
+        hot.alter('remove_row', 3, 1);
+
+        expect(hot.getData().length).toEqual(4);
```
```suggestion
        expect(hot.countRows()).toBe(4);
```

budnix commented on src/plugins/formulas/__tests__/formulas.spec.js:
```diff
@@ -2098,6 +2098,197 @@ describe('Formulas general', () => {
+
+        hot.alter('remove_col', 3, 1);
+
+        expect(hot.getData().map(row => row.length)).toEqual([8, 8]);
```
```suggestion
        expect(hot.countCols()).toBe(8);
```

budnix commented on src/plugins/formulas/__tests__/formulas.spec.js:
```diff
@@ -2098,6 +2098,197 @@ describe('Formulas general', () => {
+
+        hot.alter('insert_row', 3, 1);
+
+        expect(hot.getData().length).toEqual(4);
```
```suggestion
        expect(hot.countRows()).toBe(4);
```

budnix commented on src/plugins/formulas/__tests__/formulas.spec.js:
```diff
@@ -2098,6 +2098,197 @@ describe('Formulas general', () => {
+
+        hot.alter('insert_col', 3, 1);
+
+        expect(hot.getData().map(row => row.length)).toEqual([8, 8]);
```
```suggestion
        expect(hot.countCols()).toBe(8);
```

budnix commented on src/plugins/formulas/formulas.js:
```diff
@@ -491,12 +492,13 @@ export class Formulas extends BasePlugin {
-      return false;
-    }
+  onBeforeRemoveCol(col, amount, physicalColumns) {
+    return physicalColumns.every((physicalColumn) => {
```
Keep in mind that the return value overwrites the first argument. When the engine allows removing columns or rows the `undefined` should be returned.

jansiegel replied:
Not `col`? Still, returning `true` should work just as well, right? https://github.com/handsontable/handsontable/blob/master/src/dataMap.js#L482

budnix replied:
No, if you try listen to the changes e.q 
```js
beforeRemoveCol(col) {
  console.log('beforeRemoveCol', col);
},
```
with Formulas enabled you've got `beforeRemoveCol true` in the console.

jansiegel replied:
True 👍

budnix reviewed (approved):
Thanks! Now everything is covered 👏
